pubtools/pulplib/_impl/client/client.py

```diff
@@ -239,7 +287,14 @@ def _unpack_response(cls, pulp_response):
             parsed = pulp_response.json()
         except Exception:
             # Couldn't parse as JSON?
-            # If the response was unsuccessful, raise that.
+            if pulp_response.status_code == 404 and str(pulp_response.url).endswith(
+                cls.MAINTENANCE_REPORT_URL
```

Comment:
I still very much would like to see a solution to this which doesn't involve putting maintenance mode specific logic into this generic unpack_response method.  It seems like _do_get_maintenance itself should be catching the 404 and translating it into None.

Doing it here is messy and risks problems later.  For example, you don't really know that the executor will never be used to fetch some *other* URL which happens to end with "pulp/isos/redhat-maintenance/repos.json", for a reason other than maintenance mode.  Imagine a script trying to do something like "download all files from a FileRepository", and then the error handling behavior for this one file is different for no good reason.

I think what you would need to do is something like this, in _do_get_maintenance:

```
response = self._request_executor.submit(self._do_request, method="GET", url=url)
# If we get a 404, translate it to None instead
return _map_404_to_none(response)

...

def _map_404_to_none(response):
  """Given a future, if it resolved with a 404 error then return a Future[None], else return original future"""
  out = Future()

  def handle_response(f):
    exception = f.exception()
    if hasattr(exception, 'response') and response.status_code == 404:
      # failed with 404 => returned future should return None
      out.set_result(None)
    elif exception:
      # failed with other exception => returned future should raise that exception
      out.set_exception(exception)
    else:
      # succeeded => returned future should pass through the result
      out.set_result(f.result())

  response.add_done_callback(handle_response)
  return out
```

Ideally, more-executors would have had some utility function to make this easier, but currently it doesn't, map/flatmap are only designed for the successful case.

Actually... what I wrote above is to support non-blocking style.  But I think you chose to make it blocking anyway, so couldn't you write it even simpler by making _do_get_maintenance also blocking and then simply try/except for the exception you want to handle?

Reply:
I saw it in email and had trouble finding it on github.  If you would share the code for that attempt, I could try to figure out why it didn't work, but otherwise not sure.